File: sdk/python/adrian/pii/_redactor.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 SecureAgentics

"""PairedEvent PII redactor.

Applies PII redaction to text fields within a ``PairedEvent``'s
data payload, agent context, and parent context.  Also provides
``RedactingHandler``, an ``EventHandler`` wrapper that redacts
events before forwarding them to a downstream handler.
"""

from __future__ import annotations

from copy import deepcopy
from typing import TYPE_CHECKING, Any

from adrian.format.types import (
    AgentContext,
    LlmPairData,
    PairedEvent,
    ParentContext,
    ToolPairData,
)
from adrian.pii._engine import PiiConfig, redact_text
from adrian.pii._patterns import detect

if TYPE_CHECKING:
    from adrian.hooks import EventHandler


class PiiRedactor:
    """Redacts PII from ``PairedEvent`` objects.

    Args:
        config: PII redaction configuration.
    """

    def __init__(self, config: PiiConfig | None = None) -> None:
        """Initialise with redaction config.

        Args:
            config: PII redaction settings.  Uses defaults if ``None``.
        """
        self._config = config or PiiConfig()

    def redact_event(
        self,
        event: PairedEvent,
        *,
        in_place: bool = False,
    ) -> PairedEvent:
        """Redact PII from all text fields in a PairedEvent.

        Args:
            event: The paired event to redact.
            in_place: If ``True``, mutate the event directly.  If
                ``False``, return a deep copy with redactions applied.

        Returns:
            The redacted ``PairedEvent`` (same object if in_place).
        """
        target = event if in_place else deepcopy(event)

        self._redact_agent_context(target.agent)

        if target.parent is not None:
            self._redact_parent_context(target.parent)

        if isinstance(target.data, LlmPairData):
            self._redact_llm_data(target.data)
        else:
            # Union is LlmPairData | ToolPairData; this branch is the
            # ToolPairData case.
            self._redact_tool_data(target.data)

        return target

    def event_has_pii(self, event: PairedEvent) -> bool:
        """Return True if any text field in the event contains PII.

        Used as a cheap pre-check so RedactingHandler can skip the
        deepcopy when the event is clean.
        """
        for text in self._iter_event_text(event):
            if detect(text, types=self._config.enabled_types):
                return True
        return False

    def _iter_event_text(self, event: PairedEvent) -> list[str]:
        """Collect every string field that would be scanned for PII."""
        texts: list[str] = [
            event.agent.system_prompt,
            event.agent.user_instruction,
        ]

        if event.parent is not None:
            texts.append(event.parent.system_prompt)
            texts.append(event.parent.user_instruction)

        if isinstance(event.data, LlmPairData):
            for msg in event.data.messages:
                texts.append(msg["content"])
            texts.append(event.data.output)
            for tc in event.data.tool_calls:
                _collect_strings(tc.get("args", {}), texts)
        else:
            texts.append(event.data.input)
            texts.append(event.data.output)

        return texts

    def _redact_str(self, text: str) -> str:
        """Redact PII in a single string."""
        return redact_text(text, self._config).text

    def _redact_value(self, value: Any) -> Any:  # noqa: ANN401
        """Return a new value with strings redacted recursively.

        Containers (dict, list, tuple, set, frozenset) are rebuilt;
        the input value is never mutated.  Non-string, non-container
        scalars (int, float, bool, None, ...) pass through unchanged.
        """
        match value:
            case str():
                return self._redact_str(value)
            case dict():
                return {k: self._redact_value(v) for k, v in value.items()}  # pyright: ignore[reportUnknownVariableType]
            case list():
                return [self._redact_value(v) for v in value]  # pyright: ignore[reportUnknownVariableType]
            case tuple():
                return tuple(self._redact_value(v) for v in value)  # pyright: ignore[reportUnknownVariableType, reportUnknownArgumentType]
            case frozenset():
                return frozenset(self._redact_value(v) for v in value)  # pyright: ignore[reportUnknownVariableType, reportUnknownArgumentType]
            case set():
                return {self._redact_value(v) for v in value}  # pyright: ignore[reportUnknownVariableType]
            case _:
                return value

    def _redact_llm_data(self, data: LlmPairData) -> None:
        """Redact PII fields within ``LlmPairData`` in-place."""
        for msg in data.messages:
            msg["content"] = self._redact_str(msg["content"])

        data.output = self._redact_str(data.output)

        for tc in data.tool_calls:
            args = tc["args"]

            for key, val in args.items():
                args[key] = self._redact_value(val)  # type: ignore[assignment]

    def _redact_tool_data(self, data: ToolPairData) -> None:
        """Redact PII fields within ``ToolPairData`` in-place."""
        data.input = self._redact_str(data.input)
        data.output = self._redact_str(data.output)

    def _redact_agent_context(self, ctx: AgentContext) -> None:
        """Redact system prompt and user instruction in-place."""
        ctx.system_prompt = self._redact_str(ctx.system_prompt)
        ctx.user_instruction = self._redact_str(ctx.user_instruction)

    def _redact_parent_context(self, ctx: ParentContext) -> None:
        """Redact parent context prompts in-place."""
        ctx.system_prompt = self._redact_str(ctx.system_prompt)
        ctx.user_instruction = self._redact_str(ctx.user_instruction)


def _collect_strings(value: Any, sink: list[str]) -> None:  # noqa: ANN401
    """Append every string contained in ``value`` (recursing) to ``sink``."""
    match value:
        case str():
            sink.append(value)
        case dict():
            for v in value.values():  # pyright: ignore[reportUnknownVariableType]
                _collect_strings(v, sink)
        case list() | tuple() | set() | frozenset():
            for v in value:  # pyright: ignore[reportUnknownVariableType]
                _collect_strings(v, sink)
        case _:
            pass


class RedactingHandler:
    """``EventHandler`` wrapper that redacts PII before delegation.

    Wraps an existing handler and applies PII redaction to each
    ``PairedEvent`` before passing it downstream.  Implements the
    ``EventHandler`` protocol.

    Args:
        inner: Downstream handler to forward redacted events to.
        config: PII redaction configuration.
    """

    def __init__(
        self,
        inner: EventHandler,
        config: PiiConfig | None = None,
    ) -> None:
        """Initialise with a downstream handler and config.

        Args:
            inner: The handler to forward redacted events to.
            config: PII redaction settings.
        """
        self._inner = inner
        self._redactor = PiiRedactor(config)

    async def on_paired_event(self, event: PairedEvent) -> None:
        """Redact PII (if any) and forward to the downstream handler.

        Skips the deepcopy when no PII is detected, so clean events
        cost only the scan.
        """
        if not self._redactor.event_has_pii(event):
            await self._inner.on_paired_event(event)
            return

        redacted = self._redactor.redact_event(event, in_place=False)
        await self._inner.on_paired_event(redacted)

    async def close(self) -> None:
        """Close the downstream handler."""
        await self._inner.close()

File: sdk/typescript/packages/core/src/pii/redactor.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 SecureAgentics

import type { PairedEvent } from "../format/types.js";
import type { EventHandler, JsonValue } from "../types.js";
import { redactText, type PiiConfig } from "./engine.js";
import { detect } from "./patterns.js";

export class PiiRedactor {
  private config: PiiConfig;
  constructor(config: PiiConfig = {}) {
    this.config = config;
  }

  redactEvent(event: PairedEvent, options: { inPlace?: boolean } = {}): PairedEvent {
    const target = options.inPlace ? event : structuredClone(event);
    target.agent.systemPrompt = this.redactString(target.agent.systemPrompt);
    target.agent.userInstruction = this.redactString(target.agent.userInstruction);
    if (target.parent) {
      target.parent.systemPrompt = this.redactString(target.parent.systemPrompt);
      target.parent.userInstruction = this.redactString(target.parent.userInstruction);
    }
    if (target.data.kind === "llm") {
      for (const message of target.data.messages) message.content = this.redactString(message.content);
      target.data.output = this.redactString(target.data.output);
      for (const call of target.data.toolCalls) call.args = this.redactValue(call.args) as typeof call.args;
    } else {
      target.data.input = this.redactString(target.data.input);
      target.data.output = this.redactString(target.data.output);
    }
    return target;
  }

  eventHasPii(event: PairedEvent): boolean {
    const enabledTypes = Array.isArray(this.config.enabledTypes) ? new Set(this.config.enabledTypes) : this.config.enabledTypes ?? null;
    return this.iterEventText(event).some((text) => detect(text, enabledTypes).length > 0);
  }

  private redactString(text: string): string {
    return redactText(text, this.config).text;
  }

  private redactValue(value: JsonValue): JsonValue {
    if (typeof value === "string") return this.redactString(value);
    if (Array.isArray(value)) return value.map((item) => this.redactValue(item));
    if (value && typeof value === "object") return Object.fromEntries(Object.entries(value).map(([key, val]) => [key, this.redactValue(val)]));
    return value;
  }

  private iterEventText(event: PairedEvent): string[] {
    const texts = [event.agent.systemPrompt, event.agent.userInstruction];
    if (event.parent) texts.push(event.parent.systemPrompt, event.parent.userInstruction);
    if (event.data.kind === "llm") {
      texts.push(...event.data.messages.map((msg) => msg.content), event.data.output);
      for (const call of event.data.toolCalls) collectStrings(call.args, texts);
    } else {
      texts.push(event.data.input, event.data.output);
    }
    return texts;
  }
}

export class RedactingHandler implements EventHandler {
  private inner: EventHandler;
  private redactor: PiiRedactor;
  constructor(inner: EventHandler, config: PiiConfig = {}) {
    this.inner = inner;
    this.redactor = new PiiRedactor(config);
  }

  async onPairedEvent(event: PairedEvent): Promise<void> {
    const next = this.redactor.eventHasPii(event) ? this.redactor.redactEvent(event) : event;
    await this.inner.onPairedEvent(next);
  }

  async close(): Promise<void> {
    await this.inner.close();
  }
}

function collectStrings(value: JsonValue, sink: string[]): void {
  if (typeof value === "string") sink.push(value);
  else if (Array.isArray(value)) value.forEach((item) => collectStrings(item, sink));
  else if (value && typeof value === "object") Object.values(value).forEach((item) => collectStrings(item, sink));
}
